test/BasePaintSubscription.t.sol

```diff
@@ -0,0 +1,325 @@
+
+        vm.prank(user1);
+        subscription.subscribe{value: totalCost}(subscriptions, user1);
+        
+        mockBasePaint.setToday(801);
+        
+        vm.mockCall(
+            address(mockBasePaint),
+            abi.encodeWithSelector(MockBasePaint.mint.selector, 800, 3),
+            abi.encode()
+        );
+        
+        vm.mockCall(
+            address(mockBasePaint),
+            abi.encodeWithSelector(IERC1155.safeTransferFrom.selector),
+            abi.encode()
+        );
+        
+        address[] memory addresses = new address[](1);
+        addresses[0] = user1;
+        
+        vm.prank(owner);
+        subscription.mintBasePaints(addresses);
+        
+        assertEq(subscription.balanceOf(user1, 800), 0);
+    }
+
+    function test_MintWithDifferentDiscounts() public {
+        mockBasePaint.setToday(800);
+        
+        BasePaintSubscription.Subscription[] memory subscriptions1 = new BasePaintSubscription.Subscription[](1);
+        subscriptions1[0] = BasePaintSubscription.Subscription(800, 2);
+
+        uint256 discountedPrice1 = calculateDiscountedPrice(openEditionPrice, subscription.discountPercentage());
+        uint256 totalCost1 = 2 * discountedPrice1;
+
+        vm.prank(user1);
+        subscription.subscribe{value: totalCost1}(subscriptions1, user1);
+        
+        vm.prank(owner);
+        subscription.setDiscountPercentage(10);
+        
+        BasePaintSubscription.Subscription[] memory subscriptions2 = new BasePaintSubscription.Subscription[](1);
+        subscriptions2[0] = BasePaintSubscription.Subscription(800, 3);
+
+        uint256 discountedPrice2 = calculateDiscountedPrice(openEditionPrice, subscription.discountPercentage());
+        uint256 totalCost2 = 3 * discountedPrice2;
+
+        vm.prank(user2);
+        subscription.subscribe{value: totalCost2}(subscriptions2, user2);
+        
+        assertEq(subscription.balanceOf(user1, 800), 2);
+        assertEq(subscription.balanceOf(user2, 800), 3);
+        
+        mockBasePaint.setToday(801);
+        
+        vm.mockCall(
+            address(mockBasePaint),
+            abi.encodeWithSelector(MockBasePaint.mint.selector),
+            abi.encode()
+        );
+        
+        vm.mockCall(
+            address(mockBasePaint),
+            abi.encodeWithSelector(IERC1155.safeTransferFrom.selector),
+            abi.encode()
+        );
+        
+        address[] memory addresses = new address[](2);
+        addresses[0] = user1;
+        addresses[1] = user2;
+        
+        vm.prank(owner);
+        subscription.mintBasePaints(addresses);
+        
+        assertEq(subscription.balanceOf(user1, 800), 0);
+        assertEq(subscription.balanceOf(user2, 800), 0);
+    }
+
+    function test_UpgradeContract() public {
+        BasePaintSubscription newImplementation = new BasePaintSubscription();
+        
+        vm.prank(owner);
+        subscription.upgradeToAndCall(address(newImplementation), "");
```

Comment:
this doesn't actually verify anything. Create a subclass of our impl that has an extra method, upgrade to that, make sure the extra method works